packages/api/src/store/queue.ts

```diff
@@ -0,0 +1,150 @@
+    } catch (error) {
+      console.log("error releasing pg client", error);
+    }
+  }
+
+  async listen() {
+    this.listener = await this.db.pool.connect();
+    this.listener.query(`LISTEN ${this.channel}`);
+    this.listener.on("notification", async (msg) => {
+      console.log("listener got notification: ", msg.channel);
+      await this.handleMsg(msg);
+    });
+    this.listener.on("error", (error) => {
+      console.error("Msg Queue Listener Error ", error);
+    });
+    this.listener.on("end", () => {
+      console.log("Msq queue listener ended");
+    });
+    this.listener.on("notice", (error) => {
+      console.warn("Msg Queue Listener Notice: ", error);
+    });
+  }
+
+  async handleMsg(msg: Notification) {
+    // emit it here
+    console.log("got NOTIFICATION: ", msg.channel, msg.payload);
+  }
+  // name: string;
+  // constructor({ db, schema }) {
+  //   super({ db, schema });
+  //   this.db = db;
+  //   this.schema = schema;
+  //   this.name = schema.queue;
+  // }
+
+  // get next event in queue
+  async pop(
+    processFunc: Function,
+    opts: GetOptions = { useReplica: true }
+  ): Promise<Queue> {
+    let res: QueryResult<DBLegacyObject>;
+    await this.client.query("BEGIN");
+    res = await this.client.query(
+      sql`SELECT data FROM `
+        .append(this.name)
+        .append(
+          sql` WHERE data->>'isConsumed' = 'false' AND data->>'status' = 'pending' LIMIT 1 FOR UPDATE SKIP LOCKED`.setName(
+            `${this.name}_next_event`
+          )
+        )
+    );
+
+    if (res.rowCount < 1) {
+      return null;
+    }
+
+    let originalData = res.rows[0].data;
+    // this allows us to pass a function that will process the event within the lock
+    if (processFunc) {
+      await processFunc(originalData as Queue);
+      res.rows[0].data.isConsumed = true;
+      res.rows[0].data.modifiedAt = Date.now();
+      await this.client.query(
+        sql`UPDATE queue SET data = data || ${res.rows[0].data} `.append(
+          ` WHERE id = '${res.rows[0].data.id}'`
+        )
+      );
+    } else {
+      res.rows[0].data.status = "processing";
+      res.rows[0].data.modifiedAt = Date.now();
+      console.log("id: ", res.rows);
+      await this.client.query(
+        sql`UPDATE queue SET data = data || ${res.rows[0].data} `.append(
+          ` WHERE id = '${res.rows[0].data.id}'`
+        )
+      );
+    }
+
+    await this.client.query("COMMIT;");
+    logger.debug(`MsgQueue consuming ${res.rows[0].id}`);
+    return originalData as Queue;
+  }
+
+  async emit(doc: Queue): Promise<Queue> {
+    if (!doc.status) {
+      doc.status = "pending";
+    }
+
+    try {
+      await this.db.query(
+        `INSERT INTO ${this.name} VALUES ($1, $2)`, //p
+        [doc.id, JSON.stringify(doc)] //p
+      );
+
+      if (this.listener) {
+        // emit it via notifications
+        console.log("emitting NOTIFY");
+        await this.db.query(`NOTIFY ${this.channel}, '${doc.id}';`);
```

Comment:
Ideally we should emit `NOTIFY` inside one transaction with `INSERT`
(and it doesn't matter if there is `this.listener` - it can be that some other server listen for changes)